### PR TITLE
Add support for policy based routes to net-vpc

### PR DESCRIPTION
**Pull request by sruffilli**
This PR adds support for policy based routes to net-vpc.

---

I applicable, I acknowledge that I have:
- [x] Read the [contributing guide](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md)
- [x] Ran `terraform fmt` on all modified files
- [x] Regenerated the relevant README.md files using [`tools/tfdoc.py`](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md#fabric-tools)
- [x] Made sure all relevant tests pass

### CODE REVIEW

File: modules/net-vpc/README.md [hunk 1]
```diff
@@ -455,6 +456,41 @@ module "vpc" {
 # tftest modules=5 resources=15 inventory=routes.yaml
 ```
 
+### Custom Routes
```
Comment (ludoo) [inline modules/net-vpc/README.md:459]: s/Custom/Policy Based/

File: modules/net-vpc/routes.tf [hunk 2]
```diff
@@ -108,3 +108,35 @@ resource "google_compute_route" "vpn_tunnel" {
...
+resource "google_network_connectivity_policy_based_route" "default" {
+  for_each              = var.policy_based_routes
+  project               = var.project_id
+  network               = local.network.id
+  name                  = "${var.name}-${each.key}"
+  description           = each.value.description
+  priority              = each.value.priority
+  next_hop_other_routes = each.value.use_default_routing ? "DEFAULT_ROUTING" : null
+  next_hop_ilb_ip       = each.value.use_default_routing ? null : each.value.next_hop_ilb_ip
+
```
Comment (ludoo) [inline modules/net-vpc/routes.tf:121]: nit: can you remove these newlines?

File: modules/net-vpc/variables.tf [hunk 3]
```diff
@@ -108,6 +108,50 @@ variable "peering_config" {
...
+      for r in var.policy_based_routes :
+      contains(["TCP", "UDP", "ALL", null], r.filter.ip_protocol)
+      if r.filter.ip_protocol != null
+    ])
+    error_message = "Unsupported protocol for route."
+  }
+  validation {
+    condition = alltrue([
+      for r in var.policy_based_routes :
+      r.use_default_routing != (r.next_hop_ilb_ip != null)
```
Comment (ludoo) [inline modules/net-vpc/variables.tf:142]: I can't parse this lol

**Review by ludoo — APPROVED** [review]

File: modules/net-vpc/variables.tf (hunk 3 above)
Comment (sruffilli) [reply modules/net-vpc/variables.tf:142]: 🤣 https://developer.hashicorp.com/terraform/language/expressions/operators#:~:text=Terraform%20does%20not%20have%20an%20operator%20for%20the%20%22exclusive%20OR

Comment (sruffilli) [reply modules/net-vpc/variables.tf:142]: Happy to! 
That would be
`(r.use_default_routing == true && r.next_hop_ilb_ip == null) || (r.use_default_routing && r.next_hop_ilb_ip != null)`

Would you prefer that?

File: modules/net-vpc/routes.tf [hunk 4]
```diff
@@ -108,3 +108,32 @@ resource "google_compute_route" "vpn_tunnel" {
...
+  next_hop_other_routes = each.value.use_default_routing ? "DEFAULT_ROUTING" : null
+  next_hop_ilb_ip       = each.value.use_default_routing ? null : each.value.next_hop_ilb_ip
+  filter {
+    protocol_version = "IPV4"
+    ip_protocol      = each.value.filter.ip_protocol
+    dest_range       = each.value.filter.dest_range
+    src_range        = each.value.filter.src_range
+  }
+  dynamic "virtual_machine" {
+    for_each = each.value.target.tags != null ? { 1 = 1 } : {}
```
Comment (juliocc) [inline modules/net-vpc/routes.tf:128]: we kind of standarized on `xx ? [""] : []`

File: modules/net-vpc/variables.tf (hunk 3 above)
Comment (ludoo) [reply modules/net-vpc/variables.tf:142]: or

```hcl
(
  r.use_default_routing == true ? 1 : 0
  +
  r.next_hop_ib_ip != null ? 1 : 0
) == 1
```

**Pull request merged**